staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1864,3 +1986,523 @@ func TestForgetWatcher(t *testing.T) {
+// where:
+// - false indicates the value of the param was set to "false" by a test case
+// - true  indicates the value of the param was set to "true" by a test case
+func TestGetStartResourceVersionForWatchLockedFunc(t *testing.T) {
+	listOptions := func(allowBookmarks bool, sendInitialEvents *bool, rv string) storage.ListOptions {
+		p := storage.Everything
+		p.AllowWatchBookmarks = allowBookmarks
+
+		opts := storage.ListOptions{}
+		opts.Predicate = p
+		opts.SendInitialEvents = sendInitialEvents
+		opts.ResourceVersion = rv
+		return opts
+	}
+
+	scenarios := []struct {
+		name                      string
+		opts                      storage.ListOptions
+		requiredResourceVersion   int
+		watchCacheResourceVersion int
+
+		expectedStartResourceVersion int
+	}{
+		// +-----------------+---------------------+-----------------------+
+		// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+		// +=================+=====================+=======================+
+		// | Unset           | true/false          | nil/true/false        |
+		// +-----------------+---------------------+-----------------------+
+		{
+			name:                      "RV=unset, allowWatchBookmarks=true, sendInitialEvents=nil, expectedStartRV=0",
+			opts:                      listOptions(true, nil, ""),
+			requiredResourceVersion:   100,
+			watchCacheResourceVersion: 90,
+		},
+		{
+			name:                      "RV=unset, allowWatchBookmarks=true, sendInitialEvents=true, expectedStartRV=0",
+			opts:                      listOptions(true, pointer.Bool(true), ""),
+			requiredResourceVersion:   100,
+			watchCacheResourceVersion: 90,
+		},
+		{
+			// TODO: this looks incorrect
+			name:                         "RV=unset, allowWatchBookmarks=true, sendInitialEvents=false, expectedStartRV=90",
+			opts:                         listOptions(true, pointer.Bool(false), ""),
+			requiredResourceVersion:      100,
+			watchCacheResourceVersion:    90,
+			expectedStartResourceVersion: 90,
+		},
+		{
+			name:                      "RV=unset, allowWatchBookmarks=false, sendInitialEvents=nil, expectedStartRV=0",
+			opts:                      listOptions(false, nil, ""),
+			requiredResourceVersion:   100,
+			watchCacheResourceVersion: 90,
+		},
+		{
+			name:                      "RV=unset, allowWatchBookmarks=false, sendInitialEvents=true, expectedStartRV=0",
+			opts:                      listOptions(false, pointer.Bool(true), ""),
+			requiredResourceVersion:   100,
+			watchCacheResourceVersion: 90,
+		},
+		{
+			// TODO: this looks incorrect
+			name:                         "RV=unset, allowWatchBookmarks=false, sendInitialEvents=false, expectedStartRV=90",
+			opts:                         listOptions(false, pointer.Bool(false), ""),
+			requiredResourceVersion:      100,
+			watchCacheResourceVersion:    90,
+			expectedStartResourceVersion: 90,
+		},
+		// +-----------------+---------------------+-----------------------+
```

Comment:
I actually lost track a bit.. I no longer now why we need getStartResourceVersion function at all.

Shouldn't it be as simple as:
```
if requiredRV == 0 return c.watchCacher.resourceVersion
return requiredRV
```

?